### PR TITLE
Fix deadlock in EventPipeEventDispatcher

### LINKED ISSUE #86838: Deadlock on EventSource.EventListenersLock between EventPipeEventDispatcher.StopDispatchTask() and EventSource.Initialize()
### Description

I have encountered a deadlock on `EventSource.EventListenersLock`.

The thread holding the lock is in `EventPipeEventDispatcher.StopDispatchTask()` doing `m_dispatchTask.Wait()`:

```
>	System.Private.CoreLib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 575	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3021	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.InternalWaitCore(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 2956	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 2752	C#
 	[Waiting on Async Operation, double-click or press enter to view Async Call Stacks]	
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeEventDispatcher.StopDispatchTask() Line 153	C#
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeEventDispatcher.CommitDispatchConfiguration() Line 88	C#
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventPipeEventDispatcher.SendCommand(System.Diagnostics.Tracing.EventListener eventListener, System.Diagnostics.Tracing.EventCommand command, bool enable, System.Diagnostics.Tracing.EventLevel level, System.Diagnostics.Tracing.EventKeywords matchAnyKeywords) Line 57	C#
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventListener.EnableEvents(System.Diagnostics.Tracing.EventSource eventSource, System.Diagnostics.Tracing.EventLevel level, System.Diagnostics.Tracing.EventKeywords matchAnyKeyword, System.Collections.Generic.IDictionary<string, string> arguments) Line 4076	C#
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.EventListening.DotNetEventListener.OnEventSourceCreated(object sender, System.Diagnostics.Tracing.EventSourceCreatedEventArgs e)	Unknown
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventListener.CallBackForExistingEventSources(bool addToListenersList, System.EventHandler<System.Diagnostics.Tracing.EventSourceCreatedEventArgs> callback) Line 4412	C#
 	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventListener.EventSourceCreated.add(System.EventHandler<System.Diagnostics.Tracing.EventSourceCreatedEventArgs> value) Line 3951	C#
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.EventListening.DotNetEventListener.DotNetEventListener(Prometheus.DotNetRuntime.EventListening.IEventListener eventListener, System.Diagnostics.Tracing.EventLevel level, Prometheus.DotNetRuntime.EventListening.DotNetEventListener.GlobalOptions globalOptions)	Unknown
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.DotNetRuntimeStatsCollector.CreateEventListeners.AnonymousMethod__9_0(Prometheus.DotNetRuntime.ListenerRegistration r)	Unknown
 	System.Linq.dll!System.Linq.Enumerable.SelectEnumerableIterator<Prometheus.DotNetRuntime.ListenerRegistration, Prometheus.DotNetRuntime.EventListening.DotNetEventListener>.ToArray()	Unknown
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.DotNetRuntimeStatsCollector.CreateEventListeners()	Unknown
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.DotNetRuntimeStatsCollector.DotNetRuntimeStatsCollector(Microsoft.Extensions.DependencyInjection.ServiceProvider serviceProvider, Prometheus.CollectorRegistry metricRegistry, Prometheus.DotNetRuntime.DotNetRuntimeStatsCollector.Options options)	Unknown
 	prometheus-net.DotNetRuntime.dll!Prometheus.DotNetRuntime.DotNetRuntimeStatsBuilder.Builder.StartCollecting(Prometheus.CollectorRegistry registry)	Unknown
[snip]
```

`m_dispatchTask` is meanwhile busy dispatching events and one of these dispatches has triggered code that wants to take the same lock:

image

### Reproduction Steps

I do not have an easy way to consistently reproduce the issue but I have a process dump of the deadlocked state, if it helps.

### Expected behavior

No deadlocks when doing EventSource related opterations (in this case when accessing `Activity.Current`, which appears to use EventSource internally, though the expectation is general for all EventSource related operations).

### Actual behavior

Deadlock occurs.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

Windows 11 x64.

Reproduced both with .NET 7 and .NET 8 preview 3.

### Other information

I have a process memory dump, if it helps. Ping me on Teams (sasaares).

### LINKED ISSUE #90472: EventPipe shutdown and EventPipeEventDispatcher aren't coordinated properly
If a program returns from main it will clean up EventPipe by calling ep_shutdown, but the EventPipeEventDispatcher thread is still running and may AV when it is trying to use memory owned by EventPipe's buffers.

We should either let the EventPipeEventDispatcher session run past shutdown or put in the appropriate synchronization.

### PR DESCRIPTION
**Pull request by davmason**
Fixes #86838

In EventPipeEventDispatcher we currently wait on the session to close synchronously before updating the configuration, this can lead to a deadlock if the dispatcher thread tries to call back in to EventSource apis (as described in #86838).

This change fixes the deadlock by making stopping the session asynchronous, and moving the call to update EventPipeEventDispatcher outside of the EventListenersLock in EventListener.Dispose.

This is the test program I used to recreate the deadlock. It would reliably hit the deadlock within a minute before, and can run for hours without a deadlock after

```
// See https://aka.ms/new-console-template for more information
using System.Diagnostics;
using System.Diagnostics.Tracing;
using System.Reflection.Emit;

Console.WriteLine("Hello, World!");

List<Thread> threads = new List<Thread>();
DateTime lastUpdate = DateTime.Now;

threads.Add(new Thread(() =>
{
    while (true)
    {
        RuntimeEventListener listener = new RuntimeEventListener();
        Thread.Sleep(100);
        listener.Dispose();

        lastUpdate = DateTime.Now;
        Console.Write('.');
    }
}));

threads.ForEach(x => x.Start());

while (true)
{
    // create some GC events
    GC.Collect();
    GC.Collect();

    Thread.Sleep(1000);
    if ((DateTime.Now - lastUpdate) > TimeSpan.FromSeconds(10))
    {
        Console.WriteLine("Deadlock!");
        Debugger.Launch();
        Debugger.Break();
    }
}

[EventSource(Name = "SampleEventSource")]
class SampleEventSource : EventSource
{

}

class RuntimeEventListener : EventListener
{
    protected override void OnEventSourceCreated(EventSource eventSource)
    {
        base.OnEventSourceCreated(eventSource);

        if (eventSource.Name.Equals("Microsoft-Windows-DotNETRuntime"))
        {
            EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)0x11);
        }
    }

    protected override void OnEventWritten(EventWrittenEventArgs eventData)
    {
        base.OnEventWritten(eventData);

        Console.Write("!");
        using (SampleEventSource testSource = new SampleEventSource())
        {
            testSource.Write("OneEvent");
        }
    }
}
```

As part of this fix we discovered a latent bug in EventPipeEventDispatcher as described in #90052. We are too close to releasing .net 8 to make this kind of change so the work will happen in a later release.

### CODE REVIEW

**Comment by davmason** [comment]
There is a race condition still present - since both the old task and new task use the same `m_stopDispatchTask` flag, if the new task starts running before the old one has finished shutting down it can potentially terminate early and never deliver events. I'm working on fixing it

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 1]
```diff
@@ -4075,6 +4075,12 @@ public virtual void Dispose()
                 }
                 Validate();
             }
+
+#if FEATURE_PERFTRACING
+            // Remove the listener from the EventPipe dispatcher. EventCommand.Update with enable==false removes it.
+            // Have to send outside of EventListenersLock, or we risk deadlocks.
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:4081]: Is this comment true any longer? I'd expect you can call SendCommand() with the EventListenerLock held and it produces no deadlock.

**Review by noahfalk — APPROVED** [review]
👍

**Comment by davmason** [comment]
There is a latent issue where ep_shutdown races with the EventPipeEventDispatcher thread, filed #90472 to track.

**Comment by davmason** [comment]
The changes in this PR made the shutdown race much more prevalent, I added changes to address it. This PR now fixes #90472 as well.

**Comment by davmason** [comment]
Test failures are all unrelated, merging.

**Pull request merged**